pkgs/development/haskell-modules/lib/compose.nix

```
@@ -453,20 +440,22 @@ rec {
   });
 
   /*
-    Modify a Haskell package to add shell completion scripts for the
-    given executables produced by it. These completion scripts will be
-    picked up automatically if the resulting derivation is installed,
-    e.g. by `nix-env -i`.
-
-    Invocation:
-      generateOptparseApplicativeCompletions commands pkg
-
-
-     commands: name of an executable
-          pkg: Haskell package that builds the executables
+    Retained for backwards compatibility.
+    Use haskell.packages.*.generateOptparseApplicativeCompletions
+    which is cross aware instead.
   */
   generateOptparseApplicativeCompletions = commands: pkg:
-    pkgs.lib.foldr generateOptparseApplicativeCompletion pkg commands;
+    lib.warn "haskellLib.generateOptparseApplicativeCompletions is deprecated in favor of haskellPackages.generateOptparseApplicativeCompletions. Please change ${pkg.name} to use the latter and make sure it uses its matching haskell.packages set!"
```

```suggestion
    lib.warnIf (lib.isInOldestRelease 2211) "haskellLib.generateOptparseApplicativeCompletions is deprecated in favor of haskellPackages.generateOptparseApplicativeCompletions. Please change ${pkg.name} to use the latter and make sure it uses its matching haskell.packages set!"
```

We should give third-party projects a bit of time before warning, so that they aren't asked to apply the change before all supported Nixpkgs releases support the new method.

---

Side note: The naming of this function is so confusing, it took me like 5 minutes to figure out what it means.

---

Not sure which one you mean, but I wrote both and I agree. I'm open to suggestions.